crates/transaction-pool/src/batcher.rs

```diff
@@ -98,23 +170,51 @@ where
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         let mut this = self.project();
 
-        loop {
-            // Drain all available requests from the receiver
-            ready!(this.request_rx.poll_recv_many(cx, this.buf, *this.max_batch_size));
-
-            if !this.buf.is_empty() {
-                let batch = std::mem::take(this.buf);
-                let pool = this.pool.clone();
-                tokio::spawn(async move {
-                    Self::process_batch(&pool, batch).await;
-                });
-                this.buf.reserve(1);
-
-                continue;
-            }
-
-            // No requests available, return Pending to wait for more
-            return Poll::Pending;
+        match this.interval.as_mut().as_pin_mut() {
+            // Immediate mode (timeout = 0): zero-cost path, original behavior
+            None => loop {
+                match this.request_rx.as_mut().poll_recv_many(cx, this.buf, *this.max_batch_size) {
+                    Poll::Ready(0) => return Poll::Ready(()), // Channel closed
+                    Poll::Ready(_) => {
+                        Self::spawn_batch(this.pool, this.buf);
+                        this.buf.reserve(*this.max_batch_size);
+                        // continue to check for more requests
+                    }
+                    Poll::Pending => return Poll::Pending,
+                }
+            },
```

Comment:
all of this looks a bit too complex

how should this behave exactly?

can we not use just the interval to yield pending if it's not ready yet?

Reply:
Fixed, thanks for review